{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Primer design for HT expression of CAZymes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "<a target=\"_blank\" href=\"https://colab.research.google.com/github/pydna-group/pydna/blob/master/docs/notebooks/Example_HT_cazyme_primer_design.ipynb\">\n",
    "  <img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/>\n",
    "</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "# Install pydna (only when running on Colab)\n",
    "import sys\n",
    "if 'google.colab' in sys.modules:\n",
    "    %pip install pydna[clipboard,download,express,gel] teemi\n",
    "    !curl -LO \"https://github.com/pydna-group/pydna/raw/master/docs/notebooks/GeneByLocusTag_CAZymes.fasta\"\n",
    "    !curl -LO \"https://github.com/pydna-group/pydna/raw/master/docs/notebooks/GeneByLocusTag_prot_CAZymes.fasta\"\n",
    "    !curl -LO \"https://github.com/pydna-group/pydna/raw/master/docs/notebooks/SignalP_output_CAZymes.gff3\"\n",
    "\n",
    "#pydna-5.5.3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3",
   "metadata": {},
   "source": [
    "> ⚠️ This notebook uses the extra dependency `teemi`. Run it in google Colab, or in an environment where you install `teemi` as well as `pydna`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "Shamelessly adapeted from: https://github.com/hiyama341/cazyme_primer_design/blob/main/notebooks/00-cazyme_primer_design.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5",
   "metadata": {},
   "source": [
    "In this notebook, we design a high-throughput CAZyme library by automating key steps in the primer design workflow. Specifically, we:\n",
    "\n",
    "1. Read nucleotide sequences from a text file\n",
    "\n",
    "2. Translate sequences to protein\n",
    "\n",
    "3. Import SignalP output to identify signal peptides\n",
    "\n",
    "4. Remove signal peptides from the native sequences\n",
    "\n",
    "5. Generate primers for downstream applications\n",
    "\n",
    "This approach enables efficient processing of large sequence datasets and systematic design of primers for CAZymes without Signal peptides. In this example we have 11 targets but in principle there is not a limit. Happy bioengineering. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "Functions for the notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from typing import List, Dict, Union\n",
    "import requests\n",
    "import json\n",
    "from teemi.design.combinatorial_design import simple_amplicon_maker\n",
    "from pydna.design import primer_design\n",
    "\n",
    "def open_gff3_files(path: str = \"\") -> List[List[str]]:\n",
    "    \"\"\"\n",
    "    Opens and reads a GFF3 file and returns its contents as a list of lists.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    path: str\n",
    "        The path to the GFF3 file.\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    List[List[str]]\n",
    "        A list of lists containing the contents of the GFF3 file.\n",
    "    \"\"\"\n",
    "    with open(path, \"r\") as infile:\n",
    "        LINES = []\n",
    "        for line in infile:\n",
    "            LINES.append(line[:].split(\"\\t\"))\n",
    "        LINES = LINES[1:]\n",
    "\n",
    "    return LINES\n",
    "\n",
    "\n",
    "def tidy_up_gff(lst_of_gff: list) -> list:\n",
    "    \"\"\"\n",
    "    This function takes a list of GFF lines and returns a list of dictionaries,\n",
    "    with each dictionary containing information on the signal peptides in the GFF file.\n",
    "\n",
    "    Parameters:\n",
    "    lst_of_gff (list): A list of GFF lines.\n",
    "\n",
    "    Returns:\n",
    "    list_of_peptides (list): A list of dictionaries, with each dictionary containing information on the signal peptides in the GFF file.\n",
    "    \"\"\"\n",
    "    signal_peptides = {}\n",
    "    list_of_peptides = []\n",
    "\n",
    "    for peptide in lst_of_gff:\n",
    "        signal_peptides[\"gene\"] = peptide[0][:19]\n",
    "        signal_peptides[\"start_pos\"] = int(peptide[3]) - 1\n",
    "        signal_peptides[\"end_pos\"] = int(peptide[4]) + 1\n",
    "        signal_peptides[\"signal_peptide_likelyhood\"] = peptide[5]\n",
    "        list_of_peptides.append(signal_peptides)\n",
    "        signal_peptides = {\n",
    "            \"gene\": \"\",\n",
    "            \"start_pos\": \"\",\n",
    "            \"end_pos\": \"\",\n",
    "            \"signal_peptide_likelyhood\": \"\",\n",
    "        }\n",
    "\n",
    "    return list_of_peptides\n",
    "\n",
    "\n",
    "def dict_of_signal_peptides(path: str = \"\") -> List[Dict[str, Union[str, int]]]:\n",
    "    \"\"\"\n",
    "    Given a path to a GFF3 file, returns a list of dictionaries with information on signal peptides.\n",
    "\n",
    "    Args:\n",
    "        path (str): Path to the GFF3 file. Default is an empty string.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of dictionaries where each dictionary contains the following keys:\n",
    "            - 'gene' (str): Gene name of the signal peptide.\n",
    "            - 'start_pos' (int): Start position of the signal peptide in the protein sequence.\n",
    "            - 'end_pos' (int): End position of the signal peptide in the protein sequence.\n",
    "            - 'signal_peptide_likelyhood' (str): The likelihood of the sequence being a signal peptide.\n",
    "    \"\"\"\n",
    "    gff = open_gff3_files(path)\n",
    "    dict_of_signal_peptides = tidy_up_gff(gff)\n",
    "    return dict_of_signal_peptides\n",
    "\n",
    "\n",
    "def read_gff_to_pd(path: str = \"\") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a GFF3 file and returns a pandas DataFrame with columns 'gene', 'start_pos', 'end_pos',\n",
    "    and 'signal_peptide_likelyhood'.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    path : str\n",
    "        The path to the GFF3 file.\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    df : pandas.DataFrame\n",
    "        A DataFrame with columns 'gene', 'start_pos', 'end_pos', and 'signal_peptide_likelyhood'.\n",
    "    \"\"\"\n",
    "\n",
    "    gff = open_gff3_files(path)\n",
    "    dict_of_signal_peptides = tidy_up_gff(gff)\n",
    "    df = pd.DataFrame.from_records(dict_of_signal_peptides)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def primer_ta_neb(primer1, primer2, conc=0.4, prodcode=\"phusion-0\"):\n",
    "    \"\"\"Calculates primer pair melting temp TA,  from NEB.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    primer1 : str\n",
    "        first primer to be used for finding the optimal ta\n",
    "    primer2 : str\n",
    "        second primer to be used for finding the optimal ta\n",
    "    conc : float\n",
    "    prodcode : str\n",
    "        find product codes on nebswebsite: https://tmapi.neb.com/docs/productcodes\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    ta : int\n",
    "        primer pair annealing temp\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    url = \"https://tmapi.neb.com/tm/batch\"\n",
    "    seqpairs = [[primer1, primer2]]\n",
    "\n",
    "    input = {\"seqpairs\": seqpairs, \"conc\": conc, \"prodcode\": prodcode}\n",
    "    headers = {\"content-type\": \"application/json\"}\n",
    "    res = requests.post(url, data=json.dumps(input), headers=headers)\n",
    "\n",
    "    r = json.loads(res.content)\n",
    "\n",
    "    if r[\"success\"]:\n",
    "        for row in r[\"data\"]:\n",
    "            return row[\"ta\"]\n",
    "\n",
    "    else:\n",
    "        print(\"request failed\")\n",
    "        print(r[\"error\"][0])\n",
    "\n",
    "\n",
    "def primer_tm_neb1(primer, conc=0.4, prodcode=\"phusion-0\"):\n",
    "    \"\"\"Calculates a single primers melting temp from NEB.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    primer1 : str\n",
    "    conc : float\n",
    "    prodcode : str\n",
    "        find product codes on nebswebsite: https://tmapi.neb.com/docs/productcodes\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    tm : int\n",
    "        primer melting temperature\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    url = \"https://tmapi.neb.com/tm/batch\"\n",
    "    seqpairs = [[primer]]\n",
    "\n",
    "    input = {\"seqpairs\": seqpairs, \"conc\": conc, \"prodcode\": prodcode}\n",
    "    headers = {\"content-type\": \"application/json\"}\n",
    "    res = requests.post(url, data=json.dumps(input), headers=headers)\n",
    "\n",
    "    r = json.loads(res.content)\n",
    "\n",
    "    if r[\"success\"]:\n",
    "        for row in r[\"data\"]:\n",
    "            return row[\"tm1\"]\n",
    "    else:\n",
    "        print(\"request failed\")\n",
    "        print(r[\"error\"][0])\n",
    "        \n",
    "        \n",
    "def make_amplicons(\n",
    "    list_of_amplicons: list, target_tm=58, limit=10, tm_function=primer_tm_neb1):\n",
    "    \"\"\"Generates pydna.amplicons which contains primers with a target temperature.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    list_of_amplicons : list\n",
    "        list of pydna.Dseqrecords\n",
    "    target_tm : int\n",
    "        representing the target melting temperature for the primers (default=55)\n",
    "    limit: int\n",
    "        representing the minimum primer size (default=5)\n",
    "    tm_function : function\n",
    "        for calculating primer melting temperature (default=primer_tm_neb)\n",
    "\n",
    "    Returns:\n",
    "    amplicons: list\n",
    "        list of amplicon objects with designed primer sequences\n",
    "    \"\"\"\n",
    "    amplicons = []\n",
    "    for i in range(len(list_of_amplicons)):\n",
    "        amplicon = primer_design(\n",
    "            list_of_amplicons[i],\n",
    "            target_tm=target_tm,\n",
    "            limit=limit,\n",
    "            tm_function=tm_function,\n",
    "        )\n",
    "\n",
    "        amplicons.append(amplicon)\n",
    "\n",
    "    return amplicons\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8",
   "metadata": {},
   "source": [
    "# Steps\n",
    "\n",
    "I would like:\n",
    "1) Read sequences from a text file\n",
    "2) Translate sequences\n",
    "3) Read in SignalP output\n",
    "4) Remove signal peptides from native sequences\n",
    "5) Generate primers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9",
   "metadata": {},
   "source": [
    "# 1 - Read sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: AO090005001389\n",
      "Name: AO090005001389\n",
      "Description: AO090005001389\n",
      "Number of features: 0\n",
      "Seq('ATGGGAGCTCCCCGGCTTCGGATCAAAGGCGCAACCTTCAAAGACCCAAATAAT...TAG')\n",
      "ID: AO090038000444\n",
      "Name: AO090038000444\n",
      "Description: AO090038000444\n",
      "Number of features: 0\n",
      "Seq('ATGCGTTCCTTGTCGTCTATCGCACTCCTGTCTGTGGTGGGAGCTGCGTCTGCC...TAA')\n",
      "ID: AO090012000917\n",
      "Name: AO090012000917\n",
      "Description: AO090012000917\n",
      "Number of features: 0\n",
      "Seq('ATGCGGGCGTCACTACCGTTTCTTACTGCCTTGGGATGTATTCCAGCCGCTTTG...TGA')\n",
      "ID: AO090005001553\n",
      "Name: AO090005001553\n",
      "Description: AO090005001553\n",
      "Number of features: 0\n",
      "Seq('ATGAAATTCCGTAACCTTTTCTTTGCTGCCGTAGCTGGCTCTGCGGTTGCTGCT...TAG')\n",
      "ID: AO090011000715\n",
      "Name: AO090011000715\n",
      "Description: AO090011000715\n",
      "Number of features: 0\n",
      "Seq('ATGAGAATCGGCAACTTGATCGTGGCTGCAAGTGCTGCAAGCCTGGTGCATGCG...TAA')\n",
      "ID: AO090011000757\n",
      "Name: AO090011000757\n",
      "Description: AO090011000757\n",
      "Number of features: 0\n",
      "Seq('ATGAAAGTCACCAGGCTAGCGGTTCTGAATACCCTGGCAACCCTAACTGTTGCC...TAG')\n",
      "ID: AO090003000990\n",
      "Name: AO090003000990\n",
      "Description: AO090003000990\n",
      "Number of features: 0\n",
      "Seq('ATGTTTATTAAACTTTTAAACAAAGCGCTTCTAGTGCTTGGTCTGCTGTCAGCT...TAA')\n",
      "ID: AO090009000373\n",
      "Name: AO090009000373\n",
      "Description: AO090009000373\n",
      "Number of features: 0\n",
      "Seq('ATGACTGACCCAAACTATACACCACCAACCGTGGAGGATATATTTCGCTACCGA...TGA')\n",
      "ID: AO090012000006\n",
      "Name: AO090012000006\n",
      "Description: AO090012000006\n",
      "Number of features: 0\n",
      "Seq('ATGGCAATGCTGTTCACTCCCATTGCGGCTGCATCACTGCTAGCGGTGGTGGGC...TAA')\n",
      "ID: AO090001000604\n",
      "Name: AO090001000604\n",
      "Description: AO090001000604\n",
      "Number of features: 0\n",
      "Seq('ATGCCGAGTCACTCGCGCAGTCGGGACCGCTATCGCAGCGAGCGGGATCCGAGC...TAA')\n",
      "ID: AO090010000122\n",
      "Name: AO090010000122\n",
      "Description: AO090010000122\n",
      "Number of features: 0\n",
      "Seq('ATGGAGCTCTCTTCTGGATCAAGAGCGGAGCATCGCCCGATCCACGCGTGGGCC...TGA')\n",
      "ID: AO090003001590\n",
      "Name: AO090003001590\n",
      "Description: AO090003001590\n",
      "Number of features: 0\n",
      "Seq('ATGCAACCATTACGGCTTTTGGCTCTCACAGCCATTCTTAAGGGCGCCTGGGCC...TGA')\n"
     ]
    }
   ],
   "source": [
    "sequences = list()\n",
    "for sequence in SeqIO.parse('../workflow_data/GeneByLocusTag_CAZymes.fasta',format='fasta'):\n",
    "    print(sequence)\n",
    "    sequences.append(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: AO090005001389\n",
      "Name: AO090005001389\n",
      "Description: AO090005001389\n",
      "Number of features: 0\n",
      "Seq('ATGGGAGCTCCCCGGCTTCGGATCAAAGGCGCAACCTTCAAAGACCCAAATAAT...TAG')\n"
     ]
    }
   ],
   "source": [
    "print(sequences[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13",
   "metadata": {},
   "source": [
    "# 2 - Translate sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: AO090005001389\n",
      "Name: AO090005001389\n",
      "Description: AO090005001389\n",
      "Number of features: 0\n",
      "Seq('MGAPRLRIKGATFKDPNNREITLRGINVAGESKYPKSPDTPSYVPDKFFETDDV...TIM')\n",
      "ID: AO090038000444\n",
      "Name: AO090038000444\n",
      "Description: AO090038000444\n",
      "Number of features: 0\n",
      "Seq('MRSLSSIALLSVVGAASAQAGPWAQCGGKSFSGSSECASGWKCQELNEWFSQCV...NKN')\n",
      "ID: AO090012000917\n",
      "Name: AO090012000917\n",
      "Description: AO090012000917\n",
      "Number of features: 0\n",
      "Seq('MRASLPFLTALGCIPAALAAPHPRVQSPEYVNWTTFKANGVNLGGWLVQESTID...DTK')\n",
      "ID: AO090005001553\n",
      "Name: AO090005001553\n",
      "Description: AO090005001553\n",
      "Number of features: 0\n",
      "Seq('MKFRNLFFAAVAGSAVAAPLAKEQKKRDSVFQWIGANESGAEFGENNLPGVWGT...YLG')\n",
      "ID: AO090011000715\n",
      "Name: AO090011000715\n",
      "Description: AO090011000715\n",
      "Number of features: 0\n",
      "Seq('MRIGNLIVAASAASLVHAYPTRDIKKRGSGFTWVGVSESGAEFGSSIPGTLGTD...QCL')\n",
      "ID: AO090011000757\n",
      "Name: AO090011000757\n",
      "Description: AO090011000757\n",
      "Number of features: 0\n",
      "Seq('MKVTRLAVLNTLATLTVAWLPTTDKTITSSNGTDLFKASHGKIRGVNLGSQFVF...RSS')\n",
      "ID: AO090003000990\n",
      "Name: AO090003000990\n",
      "Description: AO090003000990\n",
      "Number of features: 0\n",
      "Seq('MFIKLLNKALLVLGLLSAGTQAATIRLDPRASSFDYNGEKVRGVNLGGWLVLEP...KQC')\n",
      "ID: AO090009000373\n",
      "Name: AO090009000373\n",
      "Description: AO090009000373\n",
      "Number of features: 0\n",
      "Seq('MTDPNYTPPTVEDIFRYRYQHGTNLGSMFVHGPWLSDGASSSDSGGSRELEEVK...SVH')\n",
      "ID: AO090012000006\n",
      "Name: AO090012000006\n",
      "Description: AO090012000006\n",
      "Number of features: 0\n",
      "Seq('MAMLFTPIAAASLLAVVGTQPTGVFAQDAEGWYKAHPGMSRIKDVNQDTHQIVD...TRK')\n",
      "ID: AO090001000604\n",
      "Name: AO090001000604\n",
      "Description: AO090001000604\n",
      "Number of features: 0\n",
      "Seq('MPSHSRSRDRYRSERDPSRRYREVYDDDDDDDFDYHPRERRRYRRDDYQHDIRS...EYY')\n",
      "ID: AO090010000122\n",
      "Name: AO090010000122\n",
      "Description: AO090010000122\n",
      "Number of features: 0\n",
      "Seq('MELSSGSRAEHRPIHAWAQNVYQGRISPESRFYKNSRILKSELSFRNRQRTANN...SRK')\n",
      "ID: AO090003001590\n",
      "Name: AO090003001590\n",
      "Description: AO090003001590\n",
      "Number of features: 0\n",
      "Seq('MQPLRLLALTAILKGAWALSANCTGSFDAISASDFVANINPGWNLGNSLDATPN...LTV')\n"
     ]
    }
   ],
   "source": [
    "proteins = list()\n",
    "for sequence in SeqIO.parse('../workflow_data/GeneByLocusTag_prot_CAZymes.fasta',format='fasta'):\n",
    "    print(sequence)\n",
    "    proteins.append(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15",
   "metadata": {},
   "source": [
    "# 3 - Read in SignalP output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>signal_peptide_likelyhood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090038000444</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.9994235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090012000917</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.99983215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005001553</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>0.999823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090011000715</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>0.9998376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090011000757</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.9998413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090003000990</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0.9998205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AO090012000006</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0.99984413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AO090003001590</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.99982584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene  start_pos  end_pos signal_peptide_likelyhood\n",
       "0  AO090038000444          0       19                 0.9994235\n",
       "1  AO090012000917          0       20                0.99983215\n",
       "2  AO090005001553          0       28                  0.999823\n",
       "3  AO090011000715          0       28                 0.9998376\n",
       "4  AO090011000757          0       19                 0.9998413\n",
       "5  AO090003000990          0       23                 0.9998205\n",
       "6  AO090012000006          0       27                0.99984413\n",
       "7  AO090003001590          0       19                0.99982584"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "signal_pep = read_gff_to_pd('../workflow_data/SignalP_output_CAZymes.gff3')\n",
    "signal_pep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': ['AO090038000444',\n",
       "  'AO090012000917',\n",
       "  'AO090005001553',\n",
       "  'AO090011000715',\n",
       "  'AO090011000757',\n",
       "  'AO090003000990',\n",
       "  'AO090012000006',\n",
       "  'AO090003001590'],\n",
       " 'n_pos': [19, 20, 28, 28, 19, 23, 27, 19]}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N_pos = signal_pep['end_pos'].to_list()\n",
    "genes_with_signal_peptides = signal_pep['gene'].to_list()\n",
    "\n",
    "my_dict = {'name':genes_with_signal_peptides, 'n_pos':N_pos }\n",
    "my_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18",
   "metadata": {},
   "source": [
    "# 4 - Remove signal peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "clean_seq = list()\n",
    "for i in range(len(genes_with_signal_peptides)):\n",
    "    for seq in sequences: \n",
    "        if genes_with_signal_peptides[i] == seq.id:\n",
    "            clean_seq.append(seq[N_pos[i]*3:-3])\n",
    "            \n",
    "for seq in sequences:\n",
    "    if seq.id not in genes_with_signal_peptides:\n",
    "        clean_seq.append(seq[3:-3])\n",
    "            \n",
    "len(clean_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('GCTGGTCCTTGGGCTCAGTGTGGAGGCAAGTCCTTCTCCGGCTCATCCGAGTGT...AAC'), id='AO090038000444', name='AO090038000444', description='AO090038000444', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CCCCATCCCCGAGTGCAGAGTCCGGAGTATGTCAACTGGACAACTTTCAAAGCC...AAG'), id='AO090012000917', name='AO090012000917', description='AO090012000917', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TCAGTTTTCCAATGTTAGTCTGCCCTTCTTCTTCTTCTTCTTCTTCTTCTTTCT...GGG'), id='AO090005001553', name='AO090005001553', description='AO090005001553', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TCGGGATTCACCTGTACGTCAAGTATTGGCTCTGAAGTTACTGGATCTAACGAA...CTG'), id='AO090011000715', name='AO090011000715', description='AO090011000715', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CTTCCAACAACCGACAAGACAATCACCTCTAGTAATGGAACCGATCTCTTCAAA...AGC'), id='AO090011000757', name='AO090011000757', description='AO090011000757', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ACTATGTCAGTATAACCGTGATGTTACCCCTTTTGCTCTGCATAGTCCCTTACT...TGC'), id='AO090003000990', name='AO090003000990', description='AO090003000990', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GATGCCGAGGGTTGGTACAAGGCTCATCCCGGAATGTCGCGCATCAAGGATGTG...AAA'), id='AO090012000006', name='AO090012000006', description='AO090012000006', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('AGTGCCAATTGTACAGGGTCCTTTGACGCCATCTCGGCCTCTGATTTCGTCGCA...GTT'), id='AO090003001590', name='AO090003001590', description='AO090003001590', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GGAGCTCCCCGGCTTCGGATCAAAGGCGCAACCTTCAAAGACCCAAATAATCGA...ATG'), id='AO090005001389', name='AO090005001389', description='AO090005001389', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ACTGACCCAAACTATACACCACCAACCGTGGAGGATATATTTCGCTACCGATAT...CAT'), id='AO090009000373', name='AO090009000373', description='AO090009000373', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CCGAGTCACTCGCGCAGTCGGGACCGCTATCGCAGCGAGCGGGATCCGAGCAGA...TAT'), id='AO090001000604', name='AO090001000604', description='AO090001000604', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GAGCTCTCTTCTGGATCAAGAGCGGAGCATCGCCCGATCCACGCGTGGGCCCAG...AAG'), id='AO090010000122', name='AO090010000122', description='AO090010000122', dbxrefs=[])]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydna.dseqrecord import Dseqrecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "22",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_seq = [Dseqrecord(seq) for seq in clean_seq]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dseq(-1510)\n",
       "GCTG..GAAC\n",
       "CGAC..CTTG"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_seq[0].seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24",
   "metadata": {},
   "source": [
    "# 5 - Generate primers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "25",
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicons = make_amplicons(clean_seq,\n",
    "                           target_tm=60, # target melting temperature\n",
    "                           limit=15,  # minimum primer length\n",
    "                           tm_function = primer_tm_neb1 #calculating melting temperature \n",
    "                           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCTGGTCCTTGGGCT')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amplicons[0].forward_primer.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "27",
   "metadata": {},
   "outputs": [],
   "source": [
    "forward_primer = [str(f.forward_primer.seq) for f in amplicons]\n",
    "r_primer = [str(r.reverse_primer.seq) for r in amplicons]\n",
    "name = [str(r.name) for r in amplicons]\n",
    "gene = [str(r.id) for r in clean_seq]\n",
    "aneal_f = [primer_tm_neb1(str(r)) for r in forward_primer]\n",
    "aneal_r = [primer_tm_neb1(str(r)) for r in r_primer]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "28",
   "metadata": {},
   "outputs": [],
   "source": [
    "ta= [primer_ta_neb(str(f.forward_primer.seq),str(f.reverse_primer.seq))  for f in amplicons]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[61, 61, 62, 61, 61, 63, 61, 62, 63, 63, 62, 62]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame({'gene_name (locus_tag)': gene, \n",
    "                   'f_seq':forward_primer, \n",
    "                   'r_seq':r_primer,\n",
    "                    'f_tm': aneal_f, \n",
    "                    'r_tm': aneal_r, \n",
    "                    'ta':ta})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name (locus_tag)</th>\n",
       "      <th>f_seq</th>\n",
       "      <th>r_seq</th>\n",
       "      <th>f_tm</th>\n",
       "      <th>r_tm</th>\n",
       "      <th>ta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AO090038000444</td>\n",
       "      <td>GCTGGTCCTTGGGCT</td>\n",
       "      <td>GTTCTTGTTGACATTTTCCATATGGT</td>\n",
       "      <td>57</td>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AO090012000917</td>\n",
       "      <td>CCCCATCCCCGAGTG</td>\n",
       "      <td>CTTCGTATCATGGAATCTCGTAAG</td>\n",
       "      <td>57</td>\n",
       "      <td>58</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AO090005001553</td>\n",
       "      <td>TCAGTTTTCCAATGTTAGTCTGC</td>\n",
       "      <td>CCCCAGGTATGCCTGC</td>\n",
       "      <td>59</td>\n",
       "      <td>58</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AO090011000715</td>\n",
       "      <td>TCGGGATTCACCTGTACG</td>\n",
       "      <td>CAGGCACTGGCTGTAGT</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AO090011000757</td>\n",
       "      <td>CTTCCAACAACCGACAAGAC</td>\n",
       "      <td>GCTGCTTCGACCTTTACAG</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AO090003000990</td>\n",
       "      <td>ACTATGTCAGTATAACCGTGATGTTAC</td>\n",
       "      <td>GCACTGCTTAGGGAACTCC</td>\n",
       "      <td>60</td>\n",
       "      <td>59</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AO090012000006</td>\n",
       "      <td>GATGCCGAGGGTTGGT</td>\n",
       "      <td>TTTACGAGTGATTGTGATAGTGATGT</td>\n",
       "      <td>58</td>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AO090003001590</td>\n",
       "      <td>AGTGCCAATTGTACAGGGT</td>\n",
       "      <td>AACAGTCAAAGTAAAGTTGACCG</td>\n",
       "      <td>58</td>\n",
       "      <td>59</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AO090005001389</td>\n",
       "      <td>GGAGCTCCCCGGCTT</td>\n",
       "      <td>CATGATAGTACAGGCTCCACGTT</td>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AO090009000373</td>\n",
       "      <td>ACTGACCCAAACTATACACCAC</td>\n",
       "      <td>ATGAACACTGTAAGTGTTGATACTAACC</td>\n",
       "      <td>59</td>\n",
       "      <td>61</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AO090001000604</td>\n",
       "      <td>CCGAGTCACTCGCGC</td>\n",
       "      <td>ATAATATTCCGGTAAATCCCCGAAAC</td>\n",
       "      <td>59</td>\n",
       "      <td>61</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AO090010000122</td>\n",
       "      <td>GAGCTCTCTTCTGGATCAAGAG</td>\n",
       "      <td>CTTACGACTGTTGATGGCCG</td>\n",
       "      <td>59</td>\n",
       "      <td>60</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name (locus_tag)                        f_seq  \\\n",
       "0         AO090038000444              GCTGGTCCTTGGGCT   \n",
       "1         AO090012000917              CCCCATCCCCGAGTG   \n",
       "2         AO090005001553      TCAGTTTTCCAATGTTAGTCTGC   \n",
       "3         AO090011000715           TCGGGATTCACCTGTACG   \n",
       "4         AO090011000757         CTTCCAACAACCGACAAGAC   \n",
       "5         AO090003000990  ACTATGTCAGTATAACCGTGATGTTAC   \n",
       "6         AO090012000006             GATGCCGAGGGTTGGT   \n",
       "7         AO090003001590          AGTGCCAATTGTACAGGGT   \n",
       "8         AO090005001389              GGAGCTCCCCGGCTT   \n",
       "9         AO090009000373       ACTGACCCAAACTATACACCAC   \n",
       "10        AO090001000604              CCGAGTCACTCGCGC   \n",
       "11        AO090010000122       GAGCTCTCTTCTGGATCAAGAG   \n",
       "\n",
       "                           r_seq  f_tm  r_tm  ta  \n",
       "0     GTTCTTGTTGACATTTTCCATATGGT    57    60  61  \n",
       "1       CTTCGTATCATGGAATCTCGTAAG    57    58  61  \n",
       "2               CCCCAGGTATGCCTGC    59    58  62  \n",
       "3              CAGGCACTGGCTGTAGT    58    58  61  \n",
       "4            GCTGCTTCGACCTTTACAG    58    58  61  \n",
       "5            GCACTGCTTAGGGAACTCC    60    59  63  \n",
       "6     TTTACGAGTGATTGTGATAGTGATGT    58    60  61  \n",
       "7        AACAGTCAAAGTAAAGTTGACCG    58    59  62  \n",
       "8        CATGATAGTACAGGCTCCACGTT    60    61  63  \n",
       "9   ATGAACACTGTAAGTGTTGATACTAACC    59    61  63  \n",
       "10    ATAATATTCCGGTAAATCCCCGAAAC    59    61  62  \n",
       "11          CTTACGACTGTTGATGGCCG    59    60  62  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "32",
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_names = [f\"#{i}_{name[i]}_{list(df['gene_name (locus_tag)'])[i]}\" for i in range(len(name))]\n",
    "\n",
    "for i in range(len(amplicon_names)):\n",
    "    amplicons[i].name = amplicon_names[i]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#0_1510bp_PCR_prod_AO090038000444\n",
      "5GCTGGTCCTTGGGCT...ACCATATGGAAAATGTCAACAAGAAC3\n",
      "                   ||||||||||||||||||||||||||\n",
      "                  3TGGTATACCTTTTACAGTTGTTCTTG5\n",
      "5GCTGGTCCTTGGGCT3\n",
      " |||||||||||||||\n",
      "3CGACCAGGAACCCGA...TGGTATACCTTTTACAGTTGTTCTTG5\n",
      "\n",
      "#1_1116bp_PCR_prod_AO090012000917\n",
      "5CCCCATCCCCGAGTG...CTTACGAGATTCCATGATACGAAG3\n",
      "                   ||||||||||||||||||||||||\n",
      "                  3GAATGCTCTAAGGTACTATGCTTC5\n",
      "5CCCCATCCCCGAGTG3\n",
      " |||||||||||||||\n",
      "3GGGGTAGGGGCTCAC...GAATGCTCTAAGGTACTATGCTTC5\n",
      "\n",
      "#2_1279bp_PCR_prod_AO090005001553\n",
      "5TCAGTTTTCCAATGTTAGTCTGC...GCAGGCATACCTGGGG3\n",
      "                           ||||||||||||||||\n",
      "                          3CGTCCGTATGGACCCC5\n",
      "5TCAGTTTTCCAATGTTAGTCTGC3\n",
      " |||||||||||||||||||||||\n",
      "3AGTCAAAAGGTTACAATCAGACG...CGTCCGTATGGACCCC5\n",
      "\n",
      "#3_1600bp_PCR_prod_AO090011000715\n",
      "5TCGGGATTCACCTGTACG...ACTACAGCCAGTGCCTG3\n",
      "                      |||||||||||||||||\n",
      "                     3TGATGTCGGTCACGGAC5\n",
      "5TCGGGATTCACCTGTACG3\n",
      " ||||||||||||||||||\n",
      "3AGCCCTAAGTGGACATGC...TGATGTCGGTCACGGAC5\n",
      "\n",
      "#4_1266bp_PCR_prod_AO090011000757\n",
      "5CTTCCAACAACCGACAAGAC...CTGTAAAGGTCGAAGCAGC3\n",
      "                        |||||||||||||||||||\n",
      "                       3GACATTTCCAGCTTCGTCG5\n",
      "5CTTCCAACAACCGACAAGAC3\n",
      " ||||||||||||||||||||\n",
      "3GAAGGTTGTTGGCTGTTCTG...GACATTTCCAGCTTCGTCG5\n",
      "\n",
      "#5_1299bp_PCR_prod_AO090003000990\n",
      "5ACTATGTCAGTATAACCGTGATGTTAC...GGAGTTCCCTAAGCAGTGC3\n",
      "                               |||||||||||||||||||\n",
      "                              3CCTCAAGGGATTCGTCACG5\n",
      "5ACTATGTCAGTATAACCGTGATGTTAC3\n",
      " |||||||||||||||||||||||||||\n",
      "3TGATACAGTCATATTGGCACTACAATG...CCTCAAGGGATTCGTCACG5\n",
      "\n",
      "#6_1481bp_PCR_prod_AO090012000006\n",
      "5GATGCCGAGGGTTGGT...ACATCACTATCACAATCACTCGTAAA3\n",
      "                    ||||||||||||||||||||||||||\n",
      "                   3TGTAGTGATAGTGTTAGTGAGCATTT5\n",
      "5GATGCCGAGGGTTGGT3\n",
      " ||||||||||||||||\n",
      "3CTACGGCTCCCAACCA...TGTAGTGATAGTGTTAGTGAGCATTT5\n",
      "\n",
      "#7_1950bp_PCR_prod_AO090003001590\n",
      "5AGTGCCAATTGTACAGGGT...CGGTCAACTTTACTTTGACTGTT3\n",
      "                       |||||||||||||||||||||||\n",
      "                      3GCCAGTTGAAATGAAACTGACAA5\n",
      "5AGTGCCAATTGTACAGGGT3\n",
      " |||||||||||||||||||\n",
      "3TCACGGTTAACATGTCCCA...GCCAGTTGAAATGAAACTGACAA5\n",
      "\n",
      "#8_2465bp_PCR_prod_AO090005001389\n",
      "5GGAGCTCCCCGGCTT...AACGTGGAGCCTGTACTATCATG3\n",
      "                   |||||||||||||||||||||||\n",
      "                  3TTGCACCTCGGACATGATAGTAC5\n",
      "5GGAGCTCCCCGGCTT3\n",
      " |||||||||||||||\n",
      "3CCTCGAGGGGCCGAA...TTGCACCTCGGACATGATAGTAC5\n",
      "\n",
      "#9_1584bp_PCR_prod_AO090009000373\n",
      "5ACTGACCCAAACTATACACCAC...GGTTAGTATCAACACTTACAGTGTTCAT3\n",
      "                          ||||||||||||||||||||||||||||\n",
      "                         3CCAATCATAGTTGTGAATGTCACAAGTA5\n",
      "5ACTGACCCAAACTATACACCAC3\n",
      " ||||||||||||||||||||||\n",
      "3TGACTGGGTTTGATATGTGGTG...CCAATCATAGTTGTGAATGTCACAAGTA5\n",
      "\n",
      "#10_2616bp_PCR_prod_AO090001000604\n",
      "5CCGAGTCACTCGCGC...GTTTCGGGGATTTACCGGAATATTAT3\n",
      "                   ||||||||||||||||||||||||||\n",
      "                  3CAAAGCCCCTAAATGGCCTTATAATA5\n",
      "5CCGAGTCACTCGCGC3\n",
      " |||||||||||||||\n",
      "3GGCTCAGTGAGCGCG...CAAAGCCCCTAAATGGCCTTATAATA5\n",
      "\n",
      "#11_1442bp_PCR_prod_AO090010000122\n",
      "5GAGCTCTCTTCTGGATCAAGAG...CGGCCATCAACAGTCGTAAG3\n",
      "                          ||||||||||||||||||||\n",
      "                         3GCCGGTAGTTGTCAGCATTC5\n",
      "5GAGCTCTCTTCTGGATCAAGAG3\n",
      " ||||||||||||||||||||||\n",
      "3CTCGAGAGAAGACCTAGTTCTC...GCCGGTAGTTGTCAGCATTC5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for amplicon in amplicons:\n",
    "    print(amplicon.name)\n",
    "    print(amplicon.figure())\n",
    "    print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_teemi_3_11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
