<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Design module &mdash; teemi 0.3.0+20.g1b653f4.dirty documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Build module" href="teemi.build.html" />
    <link rel="prev" title="teemi package" href="teemi.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> teemi
          </a>
              <div class="version">
                0.3.0+20.g1b653f4.dirty
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">teemi: Literate programming can streamline bioengineering workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#features">Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#colab-notebooks">Colab notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html#id1">0.1.0 (2023-01-02)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">teemis modules</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="teemi.html">teemi package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="teemi.html#sub-modules">Sub modules</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">Design module</a></li>
<li class="toctree-l4"><a class="reference internal" href="teemi.build.html">Build module</a></li>
<li class="toctree-l4"><a class="reference internal" href="teemi.test.html">Test module</a></li>
<li class="toctree-l4"><a class="reference internal" href="teemi.learn.html">Learn module</a></li>
<li class="toctree-l4"><a class="reference internal" href="teemi.lims.html">LIMS module</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="colab_notebooks.html">Colab notebooks</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Design module</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-teemi.design.cloning">Cloning module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-teemi.design.combinatorial_design">Combinatorial library module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="teemi.build.html">Build module</a></li>
<li class="toctree-l1"><a class="reference internal" href="teemi.test.html">Test module</a></li>
<li class="toctree-l1"><a class="reference internal" href="teemi.learn.html">Learn module</a></li>
<li class="toctree-l1"><a class="reference internal" href="teemi.lims.html">LIMS module</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">teemi</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="modules.html">teemis modules</a> &raquo;</li>
          <li><a href="teemi.html">teemi package</a> &raquo;</li>
      <li>Design module</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/teemi.design.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="design-module">
<h1>Design module<a class="headerlink" href="#design-module" title="Permalink to this heading"></a></h1>
<section id="module-teemi.design.cloning">
<span id="cloning-module"></span><h2>Cloning module<a class="headerlink" href="#module-teemi.design.cloning" title="Permalink to this heading"></a></h2>
<p>Module used for cloning of microbial strains.</p>
<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.CAS9_cutting">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">CAS9_cutting</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gRNA_record</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_record</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#CAS9_cutting"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.CAS9_cutting" title="Permalink to this definition"></a></dt>
<dd><p>Simulates double-stranded-break by CAS9 given a gRNA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gRNA_record</strong> (<em>pydna.dseqrecord.</em>) – A 20 bp DNA sequence</p></li>
<li><p><strong>background_record</strong> (<em>pydna.dseqrecord.</em>) – The sequence of interest for CRISPR mediated DSB</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>1.pydna.dseqrecord.</em> – Sequence upstream of the DSB: pydna.dseqrecord.</p></li>
<li><p><em>2. pydna.dseqrecord.</em> – Sequence downstream of the DSB: pydna.dseqrecord.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.CRIPSR_knockout">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">CRIPSR_knockout</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gRNA_record</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">insertion_site</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">repair_DNA</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#CRIPSR_knockout"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.CRIPSR_knockout" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Simple version of casembler - Cuts the insertion site with</dt><dd><p>CAS9_cutting and assembles knockout with a repair template.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gRNA_record</strong> (<em>pydna.dseqrecord.</em>) – A 20 bp DNA sequence</p></li>
<li><p><strong>insertion_site</strong> (<em>pydna.dseqrecord.</em>) – The site to knock out</p></li>
<li><p><strong>repair_DNA</strong> (<em>pydna.dseqrecord.</em>) – Repair template. Typucally 90 bp or longer</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Of assembled contig after CRISPR-mediated KO</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pydna.dseqrecord.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.UPandDW">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">UPandDW</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isite_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_gRNA_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'../data/raw/gRNAtable.csv'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#UPandDW"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.UPandDW" title="Permalink to this definition"></a></dt>
<dd><p>Finds upstream and downstream sequences based on genome and site name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strain</strong> (<em>str</em>) – name of the strain eg. CENPK113-7d
(you should specify path to the chromosome)</p></li>
<li><p><strong>isite_name</strong> (<em>str</em>) – a string of the site chomosomal site you want to retrieve</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>UP_sites</strong> (<em>list</em>) – list of pydna.dseqrecord or pydna.amplicon.Amplicon</p></li>
<li><p><strong>DW_sites</strong> (<em>list</em>) – list of pydna.dseqrecord or pydna.amplicon.Amplicon</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.USER_enzyme">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">USER_enzyme</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">amplicon</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#USER_enzyme"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.USER_enzyme" title="Permalink to this definition"></a></dt>
<dd><p>Simulates digestion with USER enzyme.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>amplicon</strong> (<em>pydna.amplicon.Amplicon</em>) – An pydna.amplicon.Amplicon to with Uracil integrated</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>USER digested Dseqrecord with USER tails</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Dseqrecord</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.casembler">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">casembler</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bg_strain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gRNAs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly_limits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_benchling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#casembler"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.casembler" title="Permalink to this definition"></a></dt>
<dd><p>Simulate in vivo assembly and integration with the possibility
of printing to gb files or send it directly to benchling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bg_strain</strong> (<em>GenBank</em>) – strain of choice eg. genbank file</p></li>
<li><p><strong>site_names</strong> (<em>list</em>) – list of names e.g. [X-3, XI-3]</p></li>
<li><p><strong>gRNAs</strong> (<em>Seqrecords</em>) – list of 20 bp seqrecords e.g. [ATF1_gRNA, CroCPR_gRNA]</p></li>
<li><p><strong>parts</strong> (<em>list</em>) – list of list of parts e.g. [[ATF1_repair_template],[CPR_repair_template]]</p></li>
<li><p><strong>assembly_limits</strong> (<em>list</em>) – list of numbers of bp assembly limits e.g. [200,400]</p></li>
<li><p><strong>assembly_names</strong> (<em>list</em>) – list of names of DNA post assembly e.g.[“X_3_tADH1_P2_pPGK1”, “XI_3_UP_DW”]</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – write DNA e.g. False</p></li>
<li><p><strong>to_benchling</strong> (<em>bool</em>) – upload DNA to benchling e.g. False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>of assembled contig</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>One dseqrecord</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.extract_gRNAs">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">extract_gRNAs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#extract_gRNAs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.extract_gRNAs" title="Permalink to this definition"></a></dt>
<dd><p>Extracts gRNAs from a template.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> (<em>pydna.dseqrecord</em><em> or </em><em>pydna.amplicon.Amplicon</em>) – a plasmid or piece of DNA</p></li>
<li><p><strong>name</strong> (<em>str</em>) – a string that would include the feature name for example: gRNA</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of with the found features and their sequences</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of pydna.dseqrecord or pydna.amplicon.Amplicon</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.extract_sites">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">extract_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">templates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#extract_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.extract_sites" title="Permalink to this definition"></a></dt>
<dd><p>This function extracts the sequences from annotated sequences based
on their names</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>annotations</strong> (<em>list</em>) – list of annotations for sequences that will be extracted</p></li>
<li><p><strong>templates</strong> (<em>list of Bio.SeqRecord.SeqRecord</em>) – A list of Bio.SeqRecord.SeqRecord with SeqFeatures</p></li>
<li><p><strong>names</strong> (<em>str</em>) – name of the sequence that will be extracted</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>record</strong> – list of extracted sites</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of Bio.SeqRecord.SeqRecord</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.extract_template_amplification_sites">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">extract_template_amplification_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">templates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">terminator</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#extract_template_amplification_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.extract_template_amplification_sites" title="Permalink to this definition"></a></dt>
<dd><p>Extracts amplifications sites from a templates features</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>templates</strong> (<em>list of Bio.SeqRecord.SeqRecord</em>) – list of Bio.SeqRecord.SeqRecord objects with SeqFeatures</p></li>
<li><p><strong>names</strong> (<em>list of strings</em>) – list of strings to be extracted</p></li>
<li><p><strong>terminator</strong> (<em>str</em>) – a string with the name of upstream terminator</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>record</strong> – list of extracted elements</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of Bio.SeqRecord.SeqRecord</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.multiply_list">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">multiply_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">myList</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#multiply_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.multiply_list" title="Permalink to this definition"></a></dt>
<dd><p>Multiplies elements one by one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>myList</strong> (<em>list</em>) – list of integers to be multiplied</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>result</strong></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.nicking_enzyme">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">nicking_enzyme</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#nicking_enzyme"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.nicking_enzyme" title="Permalink to this definition"></a></dt>
<dd><p>Nt.Bbc.CI (nicking enzyme, Nicks) a vector with the
sequence ‘CGCGTG’ on watson and ‘CGCACG’ on crick strand.
:param vector: digested Dseqrecord - usually with AsiSI or similar overhang
:type vector: Dseq</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Dseq with nick - ready for USER cloning</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.plate_plot">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">plate_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#plate_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.plate_plot" title="Permalink to this definition"></a></dt>
<dd><p>Plots a 96 well plate as a pandas df.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.Dataframe</em>) – A pandas dataframe with</p></li>
<li><p><strong>value</strong> (<em>pandas dataframe column name</em>) – The name of the pandas dataframe coloumn that you want to display</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>in a 96 well plate format of the chosen column</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.Dataframe</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>1. Initialize:
Amplicon_df = {</p>
<blockquote>
<div><p>name    location        template_name   fw_name fw_location     rv_name rv_location     prow    pcol</p>
</div></blockquote>
<p>29  PCR_G8H_01      l5_A03  VminG8H_tADH1   PR_G8H_01       op4_A10 PR_G8H_02       op4_A01 A       1
25  PCR_G8H_05      l5_A07  SmusG8H_tADH1   PR_G8H_01       op4_A10 PR_G8H_06       op4_A02 A       2
21  PCR_G8H_09      l5_B02  RsepG8H_tADH1   PR_G8H_01       op4_A10 PR_G8H_10       op4_A03 A       3
17  PCR_G8H_13      l5_B06  CacuG8H_tADH1   PR_G8H_01       op4_A10 PR_G8H_14       op4_A04 A       4
13  PCR_G8H_17      l5_C01  OpumG8H_tADH1   PR_G8H_01       op4_A10 PR_G8H_18       op4_A05 A       5
}</p>
<p>2. Call the function:  # here we call the name coloumn
plate_plot(amplicon_df, ‘name’)</p>
<p>&lt;&lt;Result:
name
pcol        1       2       3       4       5       6       7       8       9       10      11      12
prow
A   PCR_G8H_01      PCR_G8H_05      PCR_G8H_09      PCR_G8H_13      PCR_G8H_17      PCR_G8H_21      PCR_G8H_25      PCR_G8H_29      PCR_G8H_33      PCR_UP_tADH1_01 PCR_PRO_01      NaN1
B   PCR_G8H_02      PCR_G8H_06      PCR_G8H_10      PCR_G8H_14      PCR_G8H_18      PCR_G8H_22      PCR_G8H_26      PCR_G8H_30      PCR_G8H_34      PCR_TRP1-DW_02  PCR_PRO_02      NaN2
C   PCR_G8H_03      PCR_G8H_07      PCR_G8H_11      PCR_G8H_15      PCR_G8H_19      PCR_G8H_23      PCR_G8H_27      PCR_G8H_31      PCR_G8H_35      PCR_TRP1-DW_01  PCR_PRO_03      NaN3
D   PCR_G8H_04      PCR_G8H_08      PCR_G8H_12      PCR_G8H_16      PCR_G8H_20      PCR_G8H_24      PCR_G8H_28      PCR_G8H_32      PCR_G8H_36      NaN4    PCR_PRO_04      NaN5
E   PCR_CPR_01      PCR_CPR_10      PCR_CPR_03      PCR_CPR_09      PCR_CPR_02      PCR_CPR_06      PCR_CPR_07      PCR_CPR_08      PCR_CPR_04      PCR_CPR_05      PCR_PRO_05      NaN6
F   PCR_CPR_11      PCR_CPR_20      PCR_CPR_13      PCR_CPR_19      PCR_CPR_12      PCR_CPR_16      PCR_CPR_17      PCR_CPR_18      PCR_CPR_14      PCR_CPR_15      PCR_PRO_06      NaN7
G   PCR_CPR_21      PCR_CPR_30      PCR_CPR_23      PCR_CPR_29      PCR_CPR_22      PCR_CPR_26      PCR_CPR_27      PCR_CPR_28      PCR_CPR_24      PCR_CPR_25      PCR_PRO_07      NaN8
H   PCR_CPR_31      PCR_CPR_40      PCR_CPR_33      PCR_CPR_39      PCR_CPR_32      PCR_CPR_36      PCR_CPR_37      PCR_CPR_38      PCR_CPR_34      PCR_CPR_35      PCR_PRO_08      NaN9</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.recs_no_duplicates">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">recs_no_duplicates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">recs_with_duplicates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/teemi/design/cloning.html#recs_no_duplicates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.recs_no_duplicates" title="Permalink to this definition"></a></dt>
<dd><p>Removes duplicate sequences from a list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>recs_with_duplicates</strong> (<em>list</em>) – list with duplicated elements</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>without duplicates</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.recs_no_duplicates_names">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">recs_no_duplicates_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">recs_with_duplicates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#recs_no_duplicates_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.recs_no_duplicates_names" title="Permalink to this definition"></a></dt>
<dd><p>Removes duplicate names from a list</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>recs_with_duplicates</strong> (<em>list</em>) – list with duplicated elements</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>without duplicates</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.remove_features_with_negative_loc">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">remove_features_with_negative_loc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#remove_features_with_negative_loc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.remove_features_with_negative_loc" title="Permalink to this definition"></a></dt>
<dd><p>Removes a SeqFeatures if negative.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record</strong> (<em>pydna.amplicon.Amplicon.</em>) – A amplicon with SeqFeature and locations</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>record</strong> – With the negative features deleted</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pydna.amplicon.Amplicon.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.remove_tuple_duplicates">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">remove_tuple_duplicates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/teemi/design/cloning.html#remove_tuple_duplicates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.remove_tuple_duplicates" title="Permalink to this definition"></a></dt>
<dd><p>Removes tuple duplicates</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>lst</strong> (<em>list</em>) – list with duplicated elements</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>without duplicates</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.cloning.seq_to_annotation">
<span class="sig-prename descclassname"><span class="pre">teemi.design.cloning.</span></span><span class="sig-name descname"><span class="pre">seq_to_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqrec_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqrec_onto</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aType</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/cloning.html#seq_to_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.cloning.seq_to_annotation" title="Permalink to this definition"></a></dt>
<dd><p>Anotate an amplicon object from another amplicon object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqrec_from</strong> (<em>str</em>) – annotation sequence that will be extracted</p></li>
<li><p><strong>seqrec_onto</strong> (<em>list of Bio.SeqRecord.SeqRecord</em>) – A list of Bio.SeqRecord.SeqRecord with SeqFeatures</p></li>
<li><p><strong>aType</strong> (<em>str</em>) – name of the sequence that will be extracted</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>record</strong> – list of extracted sites</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of Bio.SeqRecord.SeqRecord</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-teemi.design.combinatorial_design">
<span id="combinatorial-library-module"></span><h2>Combinatorial library module<a class="headerlink" href="#module-teemi.design.combinatorial_design" title="Permalink to this heading"></a></h2>
<p>This part of the design module is used for making combinatorial libraries from DNA fragments.</p>
<dl class="py class">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">DesignAssembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_of_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pad</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_of_pad</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">56.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">35</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class able to make a combinatorial library from DNA fragments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list_of_seqs</strong> (<em>list</em>) – A list of list of a constructs of choice.</p></li>
<li><p><strong>list_of_names</strong> (<em>list</em>) – A list of list of the names wanted for the construct of choice.</p></li>
<li><p><strong>pad</strong> (<em>pydna.Dseqrecord</em>) – A nucleotide sequence to be incorporated into the primers (Max is 40 bp)</p></li>
<li><p><strong>position_of_pad</strong> (<em>int</em>) – the position in the list of seqs where the pad is incorporated (zero indexed)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A powerfull class and a lot of information can be retrieved.
Such as: showing all the amplicons needed to construct a combinatorial library
with the simple method –&gt; PCR_list_to_dataframe or Primer_list_to_dataframe.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>teemi.design.combinatorial_design.DesignAssembly object</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.PCR_list">
<span class="sig-name descname"><span class="pre">PCR_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.PCR_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.PCR_list" title="Permalink to this definition"></a></dt>
<dd><p>Returns a PCR_list</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.PCR_list_to_dataframe">
<span class="sig-name descname"><span class="pre">PCR_list_to_dataframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.PCR_list_to_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.PCR_list_to_dataframe" title="Permalink to this definition"></a></dt>
<dd><p>Prints PCR_list into a pandas dataframe</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.ShowContigs">
<span class="sig-name descname"><span class="pre">ShowContigs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.ShowContigs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.ShowContigs" title="Permalink to this definition"></a></dt>
<dd><p>Returns a string of the contigs generated by the assembly</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.ShowVariantsLibDF">
<span class="sig-name descname"><span class="pre">ShowVariantsLibDF</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.ShowVariantsLibDF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.ShowVariantsLibDF" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dataframe of all the variants</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.graphical_representation_of_assemblies">
<span class="sig-name descname"><span class="pre">graphical_representation_of_assemblies</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.graphical_representation_of_assemblies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.graphical_representation_of_assemblies" title="Permalink to this definition"></a></dt>
<dd><p>Takes in the assembly object and returns graphical report of the
fragments assembled</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.primer_list">
<span class="sig-name descname"><span class="pre">primer_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.primer_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.primer_list" title="Permalink to this definition"></a></dt>
<dd><p>Return the list of transfers in human-readable format.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.primer_list_to_dataframe">
<span class="sig-name descname"><span class="pre">primer_list_to_dataframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.primer_list_to_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.primer_list_to_dataframe" title="Permalink to this definition"></a></dt>
<dd><p>Return a pandas dataframe with list of primers.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.print_PCR_list">
<span class="sig-name descname"><span class="pre">print_PCR_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.print_PCR_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.print_PCR_list" title="Permalink to this definition"></a></dt>
<dd><p>Prints PCR_list</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.DesignAssembly.print_primer_list">
<span class="sig-name descname"><span class="pre">print_primer_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#DesignAssembly.print_primer_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.DesignAssembly.print_primer_list" title="Permalink to this definition"></a></dt>
<dd><p>Return the list of transfers in human-readable format.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.assembly_maker">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">assembly_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">combinatorial_list_of_amplicons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">35</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#assembly_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.assembly_maker" title="Permalink to this definition"></a></dt>
<dd><p>Assembles Amplicons with pad and makes new overlapping primers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>combinatorial_list_of_amplicons</strong> (<em>list</em><em>[</em><em>[</em><em>pydna.amplicon.Amplicon</em><em>]</em><em>]</em>) – the list of pydna.amplicon.Amplicon that you want generate
overlapping primers for.</p></li>
<li><p><strong>overlap</strong> (<em>int</em><em> (</em><em>default set to 35</em><em>)</em>) – How many basepair overlaps</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>List_of_assemblies</strong> – amplicons that overlaps eachother with the specified overlap value.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list[[pydna.amplicon.Amplicon]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.combinatorial_list_maker">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">combinatorial_list_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">listOflist_that_is_being_made_into_all_combinations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#combinatorial_list_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.combinatorial_list_maker" title="Permalink to this definition"></a></dt>
<dd><p>Makes all possible combinations from a list of list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>listOflist_that_is_being_made_into_all_combinations</strong> (<em>list</em><em>[</em><em>list</em><em>[</em><em>any_type</em><em>]</em><em>]</em>) – can be of any type inside the list of lists</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>combinations</strong> – all possible combinations of the list of lists</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list[tuple(any_type)]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.count_unique_parts">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">count_unique_parts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_combinations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#count_unique_parts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.count_unique_parts" title="Permalink to this definition"></a></dt>
<dd><p>Iterate through the list of predictions and save new encountered parts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Dataframe containing predictions</p></li>
<li><p><strong>Returns</strong> – </p></li>
<li><p><strong>--------</strong> – </p></li>
<li><p><strong>parts_encounteres</strong> (<em>dict</em>) – A dictionary containing the unique parts encountered in ‘G8H’,’pG8H’, ‘pCPR’, ‘CPR’ columns,
total number of unique combinations encountered in ‘Sum of parts’ and total predictions
encountered in ‘Predictions’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.empty_list_maker">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">empty_list_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_sequences</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#empty_list_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.empty_list_maker" title="Permalink to this definition"></a></dt>
<dd><p>returns empty list in the length of seqs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>list_of_sequences</strong> (<em>list</em>) – could be any list with any types</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>EmptyList</strong> – an empty list with the same dimensions</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.get_primers">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">get_primers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">List_of_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combinatorial_list_of_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combinatorial_list_of_primer_tm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#get_primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.get_primers" title="Permalink to this definition"></a></dt>
<dd><p>Returns a list of ALL primers from the combinatorial library,
updates names and what they anneal to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>List_of_assemblies</strong> (<em>list</em><em>[</em><em>list</em><em>[</em><em>pydna.amplicon.Amplicon</em><em>]</em><em>]</em>) – </p></li>
<li><p><strong>combinatorial_list_of_names</strong> (<em>list</em><em>[</em><em>(</em><em>str</em><em>)</em><em>]</em>) – </p></li>
<li><p><strong>combinatorial_list_of_primer_tm</strong> (<em>list</em><em>[</em><em>(</em><em>float</em><em>, </em><em>float</em><em>)</em><em>,</em><em>..</em><em>)</em><em>...</em><em>]</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>primers</strong> – All primers that have been made for all assemblies</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list[list[[pydna.primer.Primer, pydna.primer.Primer]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.making_assembled_contigs">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">making_assembled_contigs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_assembly_objects</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#making_assembled_contigs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.making_assembled_contigs" title="Permalink to this definition"></a></dt>
<dd><p>Assembles a list of assembly object into
linear contigs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>list_of_assembly_objects</strong> (<em>list</em><em>[</em><em>pydna.assembly.Assembly</em><em>]</em>) – these objects can be assembled into contigs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>list_of_assembly_objects</strong> – list_of_assembly_objects have been assembled into contigs</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list[]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.making_assembly_objects">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">making_assembly_objects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#making_assembly_objects"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.making_assembly_objects" title="Permalink to this definition"></a></dt>
<dd><p>Assembling amplicons into assembling class that shows
fragments, limit,nodes and which algorithm that was used
for assembling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>list_of_assemblies</strong> (<em>list</em><em>[</em><em>[</em><em>pydna.amplicon.Amplicon</em><em>]</em><em>]</em>) – list of the combinatorial libarary with overlapping ends</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>list_of_assembly_objects</strong> – shows which algorithm that was used, nodes, limit and fragments</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list[pydna.assembly.Assembly]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.simple_amplicon_maker">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">simple_amplicon_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_of_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_tm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">56.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#simple_amplicon_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.simple_amplicon_maker" title="Permalink to this definition"></a></dt>
<dd><p>Creates amplicons, updates their names</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list_of_seqs</strong> (<em>list</em><em>[</em><em>list</em><em>[</em><em>pydna.dseqrecord.Dseqrecord</em><em>]</em><em>]</em>) – List of the pydna.dseqrecord import Dseqrecord elements u want to made into amplicons</p></li>
<li><p><strong>list_of_names</strong> (<em>list</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em>) – provide names for the sequences since pydna changes their names to amplicon</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>list_of_amplicons</strong> (<em>list[pydna.amplicon.Amplicon]</em>) – list with the pydna.amplicon.Amplicon objects that have been made</p></li>
<li><p><strong>list_of_amplicon_primers</strong> (<em>list[list[(pydna.seq.Seq, pydna.seq.Seq)]]</em>) – a list of all the generated primers in tuples where index0 = forward primer
and index1=reverse primer. Both are pydna.seq.Seq objects</p></li>
<li><p><strong>list_of_amplicon_primer_temps</strong> (<em>list[list[(float, float)]]</em>) – a list of melting temperatures in tuples where index0 = forward primer melting temp
and index1=reverse primer melting temp.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.systematic_names_function">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">systematic_names_function</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">List_of_list_parts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#systematic_names_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.systematic_names_function" title="Permalink to this definition"></a></dt>
<dd><p>Returns a list of list with systematic names i.e [1,1,1], [1,2,1]… etc</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>List_of_list_parts</strong> (<em>list of list</em>) – can have anny type withing the list[list[any_type]]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of tuples with the systematic names eg. [(1,1,1),(1,2,1)]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>combinatorial_list_of_indexes</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.unique_amplicons">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">unique_amplicons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_assemblies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#unique_amplicons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.unique_amplicons" title="Permalink to this definition"></a></dt>
<dd><p>Finds Unique amplicons from a list of assemblies
:param list_of_assemblies: list of the combinatorial libarary with overlapping ends
:type list_of_assemblies: list[[pydna.amplicon.Amplicon]]</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>unique_amplicons</strong> – returns a list of unique amplicons where relavant metrics
are added to the objects.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>list[pydna.amplicon.Amplicon]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teemi.design.combinatorial_design.unique_primers">
<span class="sig-prename descclassname"><span class="pre">teemi.design.combinatorial_design.</span></span><span class="sig-name descname"><span class="pre">unique_primers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">primers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_of_assemblies</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/teemi/design/combinatorial_design.html#unique_primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teemi.design.combinatorial_design.unique_primers" title="Permalink to this definition"></a></dt>
<dd><p>Finds unique primers from a list of assemblies
:param primers: a list of all the primers made for the combinatorial library
:type primers: list[list[[pydna.primer.Primer, pydna.primer.Primer]]
:param list_of_assemblies: used here to update the names of the primers
:type list_of_assemblies: list[[pydna.amplicon.Amplicon]]</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>unique_primers</strong> – Relevant metrics for the unique primers of the combinatorial library.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>list[list(ID,Anneals_to,Sequence,Annealing_temp,Length,Price(DKK))]</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="teemi.html" class="btn btn-neutral float-left" title="teemi package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="teemi.build.html" class="btn btn-neutral float-right" title="Build module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Technical University of Denmark (DTU).</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>